keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,859 @@
+*   When pods request for a specific [RunTimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/) (`spec.runtimeClassName`), we currently use node selectors to make sure they are on appropriate nodes . Instead, kube-scheduler can automatically identify nodes that possess the corresponding capability.
+
+## Design Details
+
+### API Changes
+
+ \
+Add a field `Capabilities` field as type `map[string]string` to the `Node.Status` structure. 
+
+```
+type NodeStatus struct {
+    // ... existing fields
+    // Capabilities provides a structured way to report various capabilities of the node. Keys are typically DNS-style and values are strings.
+    // +optional 
+    // +featureGate=NodeCapabilities
+    Capabilities map[string]string `json:"capabilities,omitempty"`
+}
+
+
+// Node object remains unchanged in spec, only status is modified.
+type Node struct {
+    ...
+    Status NodeStatus `json:"status,omitempty" `
+}
+
+```
+
+**Example:**
+
+```
+Capabilities:
+   feature.kubernetes.io/inPlacePodResize: "true"
+   feature.kubernetes.io/podLevelResourcesSupport: "true"
+   feature.kubernetes.io/swapSupport: "true"
+   runtime.kubernetes.io/handlers: "runc, runsc"
+   kernel.kubernetes.io/LSMConfig: "seLinux"   
+```
+
+**Capability Reporting Semantics**
+
+1. If multiple interdependent settings are required to enable a feature, they must be collapsed into a single, logical capability. This simplifies the decision-making process for control plane components like the scheduler, which should not need to understand multiple interdependent settings. 
+For example, Swap support on a node depends on several conditions: the feature gate, Kubelet configuration, and underlying kernel support. Instead of reporting these three states individually, Kubelet evaluates them and advertises a single, simple capability: kubernetes.io/feature/swapSupport: "true".
+2. A capability should only be present if `node.status.capabilities` map if the feature is enabled and functional (`true` or another descriptive string if the capability is not boolean). If a feature is disabled (e.g., a beta feature gate is explicitly set to false) or not supported at all (e.g., on an older Kubelet version or an incompatible OS), the Kubelet must omit the corresponding capability key from the map entirely. This ensures there is consistency between a node that doesn't know about the feature and one that has it disabled which simplifies the logic for control plane components.
+3. The keys and values within the `node.status.capabilities` map must adhere to the same validation rules as standard Kubernetes labels. The Kubelet validates each capability and discards any key-value pair that is invalid, logging an error.
```

Comment:
sorry, I do not understand at which stage kubelet will reject invalid key-value pairs. Are there some plugins planned for it that populate capabilitiues from the external sources? Or kubelet will validate capabilities set by other sources  directly thru API?